onnxruntime/core/providers/nuphar/mti_x86/tensor/scatter.cc

```diff
@@ -0,0 +1,210 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include "core/providers/nuphar/mti_x86/tensor/scatter.h"
+
+#include "core/codegen/mti/mti_tvm_utils.h"
+#include "core/common/common.h"
+#include <topi/detail/extern.h>
+
+namespace onnxruntime {
+namespace nuphar {
+
+// DCHECK(tvm::runtime::TypeMatch(X->dtype, kDLFloat, 32));
+
+#define STRINGIFY_1(x) #x
+#define STRINGIFY(x) STRINGIFY_1(x)
+#define GET_EXTERN_SCATTER_STR(input_type, index_type)                \
+  STRINGIFY(tvm.contrib.onnxruntime.scatter_##input_type##index_type)
+
+static int64_t DLTensorSize(const DLTensor* dl_tensor) {
+  int64_t sz = 1;
+  for (int i = 0; i < dl_tensor->ndim; ++i) {
+    sz *= dl_tensor->shape[i];
+  }
+  return sz;
+}
+
+template<class T, class Tind>
+void ScatterCommon(tvm::TVMArgs args, tvm::TVMRetValue* /*ret*/) {
+  DLTensor* input = args[0];
+  DLTensor* indices = args[1];
+  DLTensor* updates = args[2];
+  DLTensor* output = args[3];
+  int axis = args[4];
+
+  int num_dims = input->ndim;
+  DCHECK(axis < num_dims);
+
+  for (int i = 0; i < num_dims; i++) {
+    if (indices->shape[i] != updates->shape[i]) {
+      ORT_THROW("Indices vs updates dimensions differs at position=", i,
+                " ", indices->shape[i], " vs ", updates->shape[i]);
+    }
+  }
+
+  // extract indices from raw data
+  Tind* indices_data = reinterpret_cast<Tind*>(static_cast<char*>(indices->data) + indices->byte_offset);
+  int64_t indices_size = DLTensorSize(indices);
+  std::vector<Tind> indices_data_vec(indices_size);
+  int64_t axis_size = input->shape[axis];
+  for (int64_t i = 0; i < indices_size; i++) {
+    Tind idx = indices_data[i];
+    // indices can be negative values
+    if (idx >= -axis_size && idx < axis_size) {
+      indices_data_vec[i] = idx >= 0 ? idx : idx + static_cast<Tind>(axis_size);
+    } else {
+      ORT_THROW("indices element out of data bounds, idx=", idx,
+                " must be within the inclusive range [", -axis_size,
+                ",", axis_size - 1, "]");
+    }
+  }
+
+  // copy input data into output
+  int64_t input_size = DLTensorSize(input);
+  memcpy(static_cast<char*>(output->data) + output->byte_offset,
+         static_cast<char*>(input->data) + input->byte_offset,
+         input_size * input->dtype.bits / 8);
+
+  // Record the size of each input dimension, i.e. the number of
+  // elements contained by a single element of current dimension.
+  // For example, for input[3][4][5][6], dimension_sizes will be
+  // [4*5*6, 5*6, 6, 1], i.e. [120, 30, 6, 1]
+  std::vector<int64_t> dimension_sizes(num_dims, 0);
```

Comment:
better to call it stride. Besides, this seems common enough that other ops should have it too. Better to move to a common utility header.

Reply:
Done.